File: submissions/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        closeToOpen = {"(": ")", "{": "}", "[": "]"}
        stack = []

        for c in s:
            if c in closeToOpen:
                stack.append(closeToOpen[c])
            else:
                if stack and stack[-1] == c:
                    stack.pop()
                else:
                    return False
        return len(stack) == 0

File: submissions/0020-valid-parentheses/solution.ts
function isValid(s: string): boolean {
  const expectedBracket: Record<string, string> = {
    "{": "}",
    "[": "]",
    "(": ")",
  };

  const stack: string[] = [];
  for (const c of s) {
    if (c in expectedBracket) {
      stack.push(expectedBracket[c]);
    } else {
      const top = stack.pop();
      if (!top || top !== c) {
        return false;
      }
    }
  }
  return stack.length === 0;
}
